include/decoder.hpp

```diff
@@ -41,16 +42,20 @@ class Decoder {
     Decoder(unsigned int receive_port, unsigned int send_port, bool packed, std::optional<std::string> input_file,
             std::optional<std::string> output_file, bool iq_or_bit_stream,
             std::optional<unsigned int> uplink_scrambling_code,
-            std::shared_ptr<PrometheusExporter>& prometheus_exporter);
+            const std::shared_ptr<PrometheusExporter>& prometheus_exporter);
     ~Decoder();
 
     void main_loop();
 
   private:
-    std::shared_ptr<LowerMac> lower_mac_{};
-    std::shared_ptr<Reporter> reporter_{};
-    std::shared_ptr<BitStreamDecoder> bit_stream_decoder_{};
-    std::unique_ptr<IQStreamDecoder> iq_stream_decoder_{};
+    /// The worker queue for the lower mac
+    std::shared_ptr<StreamingOrderedOutputThreadPoolExecutor<LowerMac::return_type>> lower_mac_work_queue_;
```

Comment:
this is passed to both the bitsteam and iqstream decoder, since the iq steam decoder is not implemented for all cases we fall back to the first one making the code a bit of a mess there. shared_ptr is currently the easiest without doing some more refactoring there.